### ISSUE TITLE
One command for "apt-get update", "apt-get upgrade", and "apt-get dist-upgrade"

### ISSUE THREAD

**Issue by aaronfranke** [Debian/apt#7]
There really should be one command for "apt-get update", "apt-get upgrade", and "apt-get dist-upgrade". I often find myself typing all 3 or I just make an alias. However, this should be available by default. You could call it "apt-get update-all" or "apt-get upgrate" or something.

Reddit discussion: https://www.reddit.com/r/linuxshowerthoughts/comments/3n6oc3/apt_should_have_one_command_by_default_that_does/

**Comment by SuicSoft** [comment]
Post to Debian BTS

**Comment by julian-klode** [comment]
Sometimes we look here as well. I also merge understandable pull requests here :)

**Comment by steckerhalter** [comment]
`upgrade` and `dist-upgrade` are two different things. `dist-upgrade` should do the same like `upgrade` but in addition it will also upgrade packages that install new packages or remove packages. it doesn't make sense to use both.

with the new wrapper `apt` `dist-upgrade` is called `full-upgrade` which better reflects what it does. `update` being a separate command makes sense IMO but I guess there could be a flag `--update` which implies `update` when using the other commands

**Comment by aaronfranke** [comment]
Yes, I suppose this issue does need some redefinition now that "apt" is released. What about this: 

One command for "apt update" and "apt full-upgrade"

**Comment by steckerhalter** [comment]
you mean `apt upgrade` and `apt full-upgrade` that includes `update`? as I have said, I would use a flag `--update` but not make it do that by default

**Comment by beauhefley** [comment]
I created a script in /bin to automatically do this:
sudo apt-get -y update
sudo apt-get -y upgrade
I saved it as aguu (apt-get update upgrade) and gave it permissions 700

**Comment by ivanlmj** [comment]
+1: it lacks a command that could perform a full upgrade, including Kernel updates, etc., which are only provided via "apt-get dist-upgrade".

**Comment by steckerhalter** [comment]
@ivanlmj no, it does include that: `apt full-upgrade`

**Comment by ivanlmj** [comment]
yeap, indeed @steckerhalter . Thanks!

**Comment by josegonzalez** [comment]
`apk` (from alpine) has a similar feature:

```
# pulls the index before installing the application
apk add --update nginx
```

It could just be a flag on the following commands:

```
# updates the cache and calls upgrade
# exits 1 *only* on failed upgrade
apt upgrade --update 

# updates the cache and calls install
# exits 1 *only* on failed install/upgrade of package
apt install --update nginx
```

**Comment by aaronfranke** [comment]
`sudo apt full-upgrade` does everything *except* for `sudo apt update`. Perhaps this can be added to `sudo apt full-upgrade`? Considering it's called `full-upgrade` it would make sense to fully upgrade.

**Comment by julian-klode** [comment]
So I understand why people would like an `--update` flag. While it sounds simple on a surface, it is brutally complex to get right, as we would end up having to parse the command line twice: First to figure out if someone specified `--update`, and then parsing the options with options for updating enabled if `--update` was passed. We are not prepared for that.

**Comment by josegonzalez** [comment]
It's one fewer command. In many cases, people currently _always_ issue an `apt-get update` even if it's not required. We could instead check if the package is in the cache and update _only when_ it isn't. This would:

- decrease the amount of time it takes to install a package (no extra waiting step)
- decrease network usage by not hitting all the apt repos for the latest list of packages

**Comment by julian-klode** [comment]
That does not make much sense, as new packages do not usually appear in existing repositories.

**Comment by josegonzalez** [comment]
The example in this case is less a completely new package, but more of a package version.

**Comment by aaronfranke** [comment]
Is there any update on this? This is one of the oldest issues I have on GitHub, would be nice to have it solved. Is it still brutally complex to solve, or maybe someone has new ideas on how to solve it?

**Comment by steckerhalter** [comment]
hmm, or maybe a config option could specify if `update` is automatically run? then you wouldn't have the parsing difficulties and you could leave the default as it is now.

from a user point of view I think having to run `update` before upgrading packages is not so intuitive and if you forget it you end up having outdated software which could potentially lead to further issues (like missing a security update)

I don't think I know any other package manager where you would have to manually update the sources like that. yum does it automatically for example.

what do you think?

**Comment by vayulove** [comment]
I like the idea of a --update flag, but if it's even slightly time consuming or involved I would be embarrassed to even ask.  You want someone else to do a host of operations including code rearranging, compiling, testing and packaging because you don't want to take the effort of typing a second command or making a script file?  I've been using apt for many years loving and admiring how amazing it is.  I don't need a new command to make it easier for me to do a simple "apt update".  I certainly do need to have it separate as it is because there are times I want to install things with the current repositories active.

**Comment by josegonzalez** [comment]
Just a note: I did actually try contributing this but the codebase is fairly obtuse for me (though I'd love some mentorship!).

That said, there isn't anything wrong with adding the nice parts of other software to the software you love.

**Comment by steckerhalter** [comment]
> I like the idea of a --update flag, but if it's even slightly time consuming or involved I would be embarrassed to even ask. You want someone else to do a host of operations including code rearranging, compiling, testing and packaging because you don't want to take the effort of typing a second command or making a script file? I've been using apt for many years loving and admiring how amazing it is. I don't need a new command to make it easier for me to do a simple "apt update". I certainly do need to have it separate as it is because there are times I want to install things with the current repositories active.

It's not about making someone do anything. It's a suggestion to improve apt. I don't want to write a script either. I type it manually now and if this is done then I will use the new `--update` flag, and if it's not done I keep on doing it manually forever,  that's it.

And yes, software is about doing all the things you wrote, it's complex, a lot of work and so on but no one is forced to work on it, right? Most people do it because they love it. Nobody should feel there is an obligation but someone might actually wanna do it like @josegonzalez :smile: :ok_hand:

**Comment by vayulove** [comment]
I vote for not changing apt if the reason is to make it one line less. There is a reason for needing it as a separate command and the only reason for combining it is for someone's idea of convenience which is pretty minimal.

**Comment by ivanleoncz** [comment]
> I like the idea of a --update flag, but if it's even slightly time consuming or involved I would be embarrassed to even ask. You want someone else to do a host of operations including code rearranging, compiling, testing and packaging because you don't want to take the effort of typing a second command or making a script file? I've been using apt for many years loving and admiring how amazing it is. I don't need a new command to make it easier for me to do a simple "apt update". I certainly do need to have it separate as it is because there are times I want to install things with the current repositories active.

It was just a suggestion of improvement, which a couple of people are upvoting for this, but if there's a huge effort behind the improvement, which isn't worth doing, so let us forget about it and move on. But to judge as embarrassing to even ask or propose an improvement, is something not good for the purposes that we all have here, which is contribution. You see: this kind of approach, makes that other professionals from different areas or even beginners on tech, feeling bad on interacting with us. Let us not be like that, and let's just express our positions, without letting room for people feeling us as hard to deal or unkind. Peace :pray:
